test/extensions/http/cache/file_system_http_cache/file_system_http_cache_test.cc

```diff
@@ -116,12 +121,75 @@ TEST_F(FileSystemHttpCacheTestWithNoDefaultCache, InitialStatsAreSetCorrectly) {
   env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-b"), file_2_contents, true);
   cache_ = std::dynamic_pointer_cast<FileSystemHttpCache>(
       http_cache_factory_->getCache(cacheConfig(cfg), context_));
+  waitForEvictionThreadIdle();
   EXPECT_EQ(cache_->stats().size_limit_bytes_.value(), max_size);
   EXPECT_EQ(cache_->stats().size_limit_count_.value(), max_count);
   EXPECT_EQ(cache_->stats().size_bytes_.value(), file_1_contents.size() + file_2_contents.size());
   EXPECT_EQ(cache_->stats().size_count_.value(), 2);
 }
 
+TEST_F(FileSystemHttpCacheTestWithNoDefaultCache, EvictsOldestFilesUntilUnderCountLimit) {
+  const std::string file_contents = "XXXXX";
+  const uint64_t max_count = 2;
+  ConfigProto cfg = testConfig();
+  cfg.mutable_max_cache_entry_count()->set_value(max_count);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-a"), file_contents, true);
+  env_.writeStringToFileForTest(absl::StrCat(cache_path_, "cache-b"), file_contents, true);
+  {
+    // TODO(ravenblack): replace this with backdating the files when that's possible.
+    absl::Mutex mu;
+    absl::MutexLock lock(&mu);
+    auto cond = []() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu) { return false; };
+    mu.AwaitWithTimeout(absl::Condition(&cond), absl::Seconds(1));
```

Comment:
Agree, but not at the cost of mocking stat() repeatedly. :)

Believe me, I am a *very* strong opponent of hard delays in tests, I definitely don't intend to leave this in long-term.

Reply:
My preference would not be to mock stat(). My preference would be to add stat() to the File interface in envoy/filesystem/filesystem.h and provide a TimeSource to MemfileImpl, which would it would use for saving the timestamp.

This will also work well with the existing usage of simulated time in cache tests.

In any case I'm highly opposed to circumventing the attempts to block spurious usage of things that are equivalent to sleep() in tests. If we do ever decide that's the only way, we should have the code that implement sleep do so with a clear override of the policy, rather than squirreling in a substitute implementation that isn't in the lint yet.

Reply:
That matches my preference. Which is why I've been making steps in that direction in filesystem.h already. But I'd rather this PR not be blocked by the chain of PRs that will be required to get filesystem into a state where it's useable here. And I'd rather the test be able to do what it needs to do to be useful, in the meantime.

Reply:
I don't want to add use-cases in the codebase of how to deliberately work around the lint check.

Why is a series of PRs needed to add stat support the memory-based filesystem? Can't that just be one PR?

Reply:
Discussed in slack, and filed #24994 to keep track.